Replace custom k3s etcd script checks with vanilla grep checks

dereknola opened the pull request:
Background:
When https://github.com/aquasecurity/kube-bench/pull/1523 was merged, it pulled the cfg checks directly from https://github.com/rancher/security-scan/tree/master/package/cfg.

Problem:
- These checks utilizes a custom sonobuoy plugin to deploy kube-bench + [additional helper scripts](https://github.com/rancher/security-scan/tree/master/package/helper_scripts). These additional helper scripts were not accounted for when copying over the cfg. K3s utilizes kine which allows sqlite and other non etcd databases. The helper scripts allowed CIS check to pass on these other DB. These helper scripts do not exist in upstream kubebench.

Solution:
- This PR fixes the k3s 1.23, 1.24, and 1.7 benchmarks, restoring the correct etcd checks

Note: This means that running these benchmarks will only pass these checks is K3s is utilizing etcd (via the `--cluster-init` embedded etcd flag). It does not solve the original problem the custom scripts rancher implemented do. 


Validation:
- k3s version v1.29.4+k3s1, with hardening applied using https://docs.k3s.io/security/hardening-guide

main:
```
== Summary total ==
42 checks PASS
8 checks FAIL
45 checks WARN
36 checks INFO
```

this PR:
```
== Summary total ==
51 checks PASS
0 checks FAIL
44 checks WARN
36 checks INFO
```

andypitcher commented:
### Test case 1: auto-tls and peer-auto-tls set to false

```
[root@k3s-standalone]# cat /etc/rancher/k3s/config.yaml.d/50-rancher.yaml | tail -5
  "etcd-arg": [
    "auto-tls=true",
    "peer-auto-tls=true"
  ]
} 
```
**Results:**
```
[root@k3s-standalone]# ./kube-bench run --targets etcd --config-dir ./security-scan-0.2.14/package/cfg --config ./security-scan-0.2.14/package/cfg/config.yaml  --benchmark k3s-cis-1.7 --group 2 --include-test-output
[INFO] 2 Etcd Node Configuration
[INFO] 2 Etcd Node Configuration
[PASS] 2.1 Ensure that the --cert-file and --key-file arguments are set as appropriate (Automated)
[PASS] 2.2 Ensure that the --client-cert-auth argument is set to true (Automated)
[PASS] 2.3 Ensure that the --auto-tls argument is not set to true (Automated)
[PASS] 2.4 Ensure that the --peer-cert-file and --peer-key-file arguments are set as appropriate (Automated)
[PASS] 2.5 Ensure that the --peer-client-cert-auth argument is set to true (Automated)
[PASS] 2.6 Ensure that the --peer-auto-tls argument is not set to true (Automated)
[PASS] 2.7 Ensure that a unique Certificate Authority is used for etcd (Automated)

== Summary etcd ==
7 checks PASS
0 checks FAIL
0 checks WARN
0 checks INFO

== Summary total ==
7 checks PASS
0 checks FAIL
0 checks WARN
0 checks INFO
```

```
[root@k3s-standalone]# grep auto-tls /var/lib/rancher/k3s/server/db/etcd/config
auto-tls: false
peer-auto-tls: false
```

### Test case 2: auto-tls and peer-auto-tls set to true

Set configuration in /etc/rancher/k3s/config.yaml.d/50-rancher.yaml and restart k3s unit.

```
[root@k3s-standalone]# cat /etc/rancher/k3s/config.yaml.d/50-rancher.yaml | tail -5
  "etcd-arg": [
    "auto-tls=true",
    "peer-auto-tls=true"
  ]
} 
```
**Results:**

```
[root@k3s-standalone]# ./kube-bench run --targets etcd --config-dir ./security-scan-0.2.14/package/cfg --config ./security-scan-0.2.14/package/cfg/config.yaml  --benchmark k3s-cis-1.7 --group 2 --include-test-output
[INFO] 2 Etcd Node Configuration
[INFO] 2 Etcd Node Configuration
[PASS] 2.1 Ensure that the --cert-file and --key-file arguments are set as appropriate (Automated)
[PASS] 2.2 Ensure that the --client-cert-auth argument is set to true (Automated)
[FAIL] 2.3 Ensure that the --auto-tls argument is not set to true (Automated)
	 auto-tls: true
[PASS] 2.4 Ensure that the --peer-cert-file and --peer-key-file arguments are set as appropriate (Automated)
[PASS] 2.5 Ensure that the --peer-client-cert-auth argument is set to true (Automated)
[FAIL] 2.6 Ensure that the --peer-auto-tls argument is not set to true (Automated)
	 peer-auto-tls: true
[PASS] 2.7 Ensure that a unique Certificate Authority is used for etcd (Automated)

== Remediations etcd ==
2.3 Edit the etcd pod specification file /var/lib/rancher/k3s/server/db/etcd/config on the master
node and either remove the --auto-tls parameter or set it to false.
  --auto-tls=false

2.6 Edit the etcd pod specification file /var/lib/rancher/k3s/server/db/etcd/config on the master
node and either remove the --peer-auto-tls parameter or set it to false.
--peer-auto-tls=false


== Summary etcd ==
5 checks PASS
2 checks FAIL
0 checks WARN
0 checks INFO

== Summary total ==
5 checks PASS
2 checks FAIL
0 checks WARN
0 checks INFO
```
### Test case 3: auto-tls and peer-auto-tls not set (default value is false)

Set configuration in /etc/rancher/k3s/config.yaml.d/50-rancher.yaml and restart k3s unit.

```
[root@k3s-standalone]# grep auto-tls /var/lib/rancher/k3s/server/db/etcd/config
[root@k3s-standalone]# echo $?
1
```

**Results:**

```
[root@pitcher-k3s-standalone-pitcher-k3s-standalone-93f23d10-hvksw tmp]# vim /etc/rancher/k3s/config.yaml.d/50-rancher.yaml
[root@pitcher-k3s-standalone-pitcher-k3s-standalone-93f23d10-hvksw tmp]# service k3s restart
Redirecting to /bin/systemctl restart k3s.service
[root@pitcher-k3s-standalone-pitcher-k3s-standalone-93f23d10-hvksw tmp]# ./kube-bench run --targets etcd --config-dir ./security-scan-0.2.14/package/cfg --config ./security-scan-0.2.14/package/cfg/config.yaml  --benchmark k3s-cis-1.7 --group 2 --include-test-output
[INFO] 2 Etcd Node Configuration
[INFO] 2 Etcd Node Configuration
[PASS] 2.1 Ensure that the --cert-file and --key-file arguments are set as appropriate (Automated)
[PASS] 2.2 Ensure that the --client-cert-auth argument is set to true (Automated)
[FAIL] 2.3 Ensure that the --auto-tls argument is not set to true (Automated)
	 auto-tls: true
[PASS] 2.4 Ensure that the --peer-cert-file and --peer-key-file arguments are set as appropriate (Automated)
[PASS] 2.5 Ensure that the --peer-client-cert-auth argument is set to true (Automated)
[FAIL] 2.6 Ensure that the --peer-auto-tls argument is not set to true (Automated)
	 peer-auto-tls: true
[PASS] 2.7 Ensure that a unique Certificate Authority is used for etcd (Automated)

== Remediations etcd ==
2.3 Edit the etcd pod specification file /var/lib/rancher/k3s/server/db/etcd/config on the master
node and either remove the --auto-tls parameter or set it to false.
  --auto-tls=false

2.6 Edit the etcd pod specification file /var/lib/rancher/k3s/server/db/etcd/config on the master
node and either remove the --peer-auto-tls parameter or set it to false.
--peer-auto-tls=false


== Summary etcd ==
5 checks PASS
2 checks FAIL
0 checks WARN
0 checks INFO

== Summary total ==
5 checks PASS
2 checks FAIL
0 checks WARN
0 checks INFO
```

andypitcher reviewed (changes requested):
@dereknola feel free to apply the changes. Regarding the $etcddatadir, we need to incorporate it in all k3s-cis-* configmap, and append it in the kube-bench's global configmap.
With regards to the `env: ` test_items we could get rid of them since k8s environment variables are not usable with k3s (configuration are passed through etcd-arg etc).

dereknola commented:
Lint appears to be failing for unrelated reasons to this PR:

```
derek@degion:~/tmp/kube-bench$ golangci-lint run -c ./.golangci.yaml 
WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
check/controls_test.go:43:12: m.Called undefined (type *mockRunner has no field or method Called) (typecheck)
	args := m.Called(c)
	          ^
check/controls_test.go:201:10: runner.On undefined (type *mockRunner has no field or method On) (typecheck)
		runner.On("Run", controls.Groups[0].Checks[0]).Return(PASS)
		       ^
check/controls_test.go:202:10: runner.On undefined (type *mockRunner has no field or method On) (typecheck)
		runner.On("Run", controls.Groups[1].Checks[0]).Return(FAIL)
		       ^
check/controls_test.go:234:10: runner.AssertExpectations undefined (type *mockRunner has no field or method AssertExpectations) (typecheck)
		runner.AssertExpectations(t)
```

@chen-keinan Should I attempt to include fixes for this lint in this PR?

chen-keinan commented:
> Lint appears to be failing for unrelated reasons to this PR:
> 
> ```
> derek@degion:~/tmp/kube-bench$ golangci-lint run -c ./.golangci.yaml 
> WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
> WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
> check/controls_test.go:43:12: m.Called undefined (type *mockRunner has no field or method Called) (typecheck)
> 	args := m.Called(c)
> 	          ^
> check/controls_test.go:201:10: runner.On undefined (type *mockRunner has no field or method On) (typecheck)
> 		runner.On("Run", controls.Groups[0].Checks[0]).Return(PASS)
> 		       ^
> check/controls_test.go:202:10: runner.On undefined (type *mockRunner has no field or method On) (typecheck)
> 		runner.On("Run", controls.Groups[1].Checks[0]).Return(FAIL)
> 		       ^
> check/controls_test.go:234:10: runner.AssertExpectations undefined (type *mockRunner has no field or method AssertExpectations) (typecheck)
> 		runner.AssertExpectations(t)
> ```
> 
> @chen-keinan Should I attempt to include fixes for this lint in this PR?

could be related to linter version or go version change

mozillazg commented on cfg/k3s-cis-1.23/master.yaml:
```diff
@@ -736,7 +736,7 @@ groups:
       - id: 1.2.26
         text: "Ensure that the --etcd-certfile and --etcd-keyfile arguments are set as appropriate (Automated)"
-        audit: "check_for_k3s_etcd.sh 1.2.29"
+        audit: "journalctl -D /var/log/journal -u k3s | grep -m1 'Running kube-apiserver'"
```
it does not check the --etcd-certfile and --etcd-keyfile arguments

mozillazg commented on cfg/k3s-cis-1.24/master.yaml:
```diff
@@ -735,7 +735,7 @@ groups:
       - id: 1.2.26
         text: "Ensure that the --etcd-certfile and --etcd-keyfile arguments are set as appropriate (Automated)"
-        audit: "check_for_k3s_etcd.sh 1.2.29"
+        audit: "journalctl -D /var/log/journal -u k3s | grep -m1 'Running kube-apiserver'"
```
ditto

mozillazg commented on cfg/k3s-cis-1.7/master.yaml:
```diff
@@ -167,7 +167,7 @@ groups:
       - id: 1.1.11
         text: "Ensure that the etcd data directory permissions are set to 700 or more restrictive (Automated)"
-        audit: "check_for_k3s_etcd.sh 1.1.11"
+        audit: "stat -c %a /var/lib/rancher/k3s/server/db/etcd"
```
```suggestion
        audit: "stat -c %a $etcddatadir"

mozillazg commented on cfg/k3s-cis-1.7/master.yaml:
```diff
@@ -738,7 +738,7 @@ groups:
       - id: 1.2.25
         text: "Ensure that the --etcd-certfile and --etcd-keyfile arguments are set as appropriate (Automated)"
-        audit: "check_for_k3s_etcd.sh 1.2.29"
+        audit: "journalctl -D /var/log/journal -u k3s | grep 'Running kube-apiserver' | tail -n1"
```
it does not check the --etcd-certfile and --etcd-keyfile arguments

dereknola replied:
This check returns all arguments passed to the kube-apiserver. So a standard result looks like
```
kube-apiserver --admission-control-config-file=/var/lib/rancher/k3s/server/psa.yaml --advertise-address=10.10.10.100 --advertise-port=6443 --allow-privileged=true --anonymous-auth=false --api-audiences=https://kubernetes.default.svc.cluster.local,k3s --audit-log-maxage=30 --audit-log-maxbackup=10 --audit-log-maxsize=100 --audit-log-path=/var/lib/rancher/k3s/server/logs/audit.log --audit-policy-file=/var/lib/rancher/k3s/server/audit.yaml --authorization-mode=Node,RBAC --bind-address=127.0.0.1 --cert-dir=/var/lib/rancher/k3s/server/tls/temporary-certs --client-ca-file=/var/lib/rancher/k3s/server/tls/client-ca.crt --egress-selector-config-file=/var/lib/rancher/k3s/server/etc/egress-selector-config.yaml --enable-admission-plugins=NodeRestriction --enable-aggregator-routing=true --enable-bootstrap-token-auth=true --encryption-provider-config=/var/lib/rancher/k3s/server/cred/encryption-config.json --encryption-provider-config-automatic-reload=true --etcd-cafile=/var/lib/rancher/k3s/server/tls/etcd/server-ca.crt --etcd-certfile=/var/lib/rancher/k3s/server/tls/etcd/client.crt --etcd-keyfile=/var/lib/rancher/k3s/server/tls/etcd/client.key --etcd-servers=https://127.0.0.1:2379 --kubelet-certificate-authority=/var/lib/rancher/k3s/server/tls/server-ca.crt --kubelet-client-certificate=/var/lib/rancher/k3s/server/tls/client-kube-apiserver.crt --kubelet-client-key=/var/lib/rancher/k3s/server/tls/client-kube-apiserver.key --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --profiling=false --proxy-client-cert-file=/var/lib/rancher/k3s/server/tls/client-auth-proxy.crt --proxy-client-key-file=/var/lib/rancher/k3s/server/tls/client-auth-proxy.key --requestheader-allowed-names=system:auth-proxy --requestheader-client-ca-file=/var/lib/rancher/k3s/server/tls/request-header-ca.crt --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-group-headers=X-Remote-Group --requestheader-username-headers=X-Remote-User --secure-port=6444 --service-account-issuer=https://kubernetes.default.svc.cluster.local --service-account-key-file=/var/lib/rancher/k3s/server/tls/service.key --service-account-lookup=true --service-account-signing-key-file=/var/lib/rancher/k3s/server/tls/service.current.key --service-cluster-ip-range=10.43.0.0/16 --service-node-port-range=30000-32767 --storage-backend=etcd3 --tls-cert-file=/var/lib/rancher/k3s/server/tls/serving-kube-apiserver.crt --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305 --tls-private-key-file=/var/lib/rancher/k3s/server/tls/serving-kube-apiserver.key"
Apr 22 20:12:19 server-0 k3s[2397]: time="2024-04-22T20:12:19Z" level=info msg="Running kube-scheduler --authentication-kubeconfig=/var/lib/rancher/k3s/server/cred/scheduler.kubeconfig --authorization-kubeconfig=/var/lib/rancher/k3s/server/cred/scheduler.kubeconfig --bind-address=127.0.0.1 --kubeconfig=/var/lib/rancher/k3s/server/cred/scheduler.kubeconfig --profiling=false --secure-port=10259"
```

From this output the checks ensure thast --etcd-certfile and --etcd-keyfile are set.

I could grep again specifically for --etcd-certfile and --etcd-keyfile, but it would accomplise the same thing.

mozillazg reviewed (changes requested):
Thanks for your contribution! I've added some comments. Please check them when you get a chance. Thanks!

mozillazg reviewed (approved):
LGTM, Thanks for your contribution!

chen-keinan commented:
lgtm 🚀 thanks for the contributions
